jlm/llvm/opt/alias-analyses/RegionAwareModRefSummarizer.cpp

```diff
@@ -712,299 +646,619 @@ RegionAwareModRefSummarizer::CreateCallGraph(const rvsdg::RvsdgModule & rvsdgMod
   std::vector<size_t> reverseTopologicalOrder;
   auto numSCCs = util::FindStronglyConnectedComponents<size_t>(
       numCallGraphNodes,
-      GetSuccessor,
+      GetSuccessors,
       sccIndex,
       reverseTopologicalOrder);
 
   // sccIndex are distributed in a reverse topological order, so the sccIndex is used
   // when creating the list of SCCs and the functions they contain
-  Context_.SccFunctions.resize(numSCCs);
+  Context_->SccFunctions.resize(numSCCs);
   for (size_t i = 0; i < lambdaNodes.size(); i++)
   {
-    Context_.SccFunctions[sccIndex[i]].insert(lambdaNodes[i]);
-    Context_.FunctionToSccIndex[lambdaNodes[i]] = sccIndex[i];
+    Context_->SccFunctions[sccIndex[i]].insert(lambdaNodes[i]);
+    Context_->FunctionToSccIndex[lambdaNodes[i]] = sccIndex[i];
+  }
+
+  // Add edges between the SCCs for all calls
+  Context_->SccCallTargets.resize(numSCCs);
+  for (size_t i = 0; i < numCallGraphNodes; i++)
+  {
+    for (auto target : callGraphSuccessors[i].Items())
+    {
+      Context_->SccCallTargets[sccIndex[i]].insert(sccIndex[target]);
+    }
   }
 
   // Also note which SCC contains all external functions
-  Context_.ExternalNodeSccIndex = sccIndex[externalNodeIndex];
+  Context_->ExternalNodeSccIndex = sccIndex[externalNodeIndex];
+}
+
+[[nodiscard]] static const rvsdg::LambdaNode &
+GetSurroundingLambdaNode(const rvsdg::Node & node)
+{
+  auto it = &node;
+  while (it)
+  {
+    if (auto lambda = dynamic_cast<const rvsdg::LambdaNode *>(it))
+      return *lambda;
+    it = it->region()->node();
+  }
+  JLM_UNREACHABLE("node was not in a lambda");
 }
```

Comment:
This function should be in the rvsdg lambda.h or so. Also, you would like to throw a std::logic_error() here.

Reply:
I have the feeling that I have a similar function already somewhere, but I do not remember where.